File: _REPO/APRESS/gdiplus-programming-in-csharp-vb.net/CH_10/CustomWindow-vb/WindowPath.vb
Option Strict On

Imports System
Imports System.IO
Imports System.Drawing
Imports System.Drawing.Drawing2D

Public NotInheritable Class WindowPath

  Private Shared mP As GraphicsPath
  Private Shared mFileNames() As String
  Private Shared mIOerror As String

  Public Shared Function GetPath(ByVal fname As String) As GraphicsPath
    Dim BeginP As Point = Point.Empty
    Dim EndP As Point = Point.Empty
    Dim P As Point = Point.Empty
    Dim s As String
    Dim x() As String
    mP = New GraphicsPath()

    Dim sr As StreamReader = New StreamReader(fname)
    sr.BaseStream.Seek(0, SeekOrigin.Begin)
    While (sr.Peek() > -1)
      s = sr.ReadLine()
      x = s.Split(New Char() {CType(",", Char)})
      P.X = Convert.ToInt32(x(0))
      P.Y = Convert.ToInt32(x(1))

      If Point.op_Equality(BeginP, Point.Empty) Then
        BeginP = P
      Else
        EndP = P
        mP.AddLine(BeginP, EndP)
        BeginP = EndP
      End If
    End While
    mP.CloseFigure()

    Return mP
  End Function

  Public Shared Sub SavePath(ByVal fname As String, ByVal P As GraphicsPath)
    Dim fs As FileStream = New FileStream(fname, FileMode.Create)
    Dim sw As StreamWriter = New StreamWriter(fs)
    Dim p2 As PointF
    For Each p2 In P.PathPoints
      sw.Write(p2.X.ToString() + "," + p2.Y.ToString() + vbCrLf)
    Next
    sw.Close()
    GetFileNames()
  End Sub

  Public Shared ReadOnly Property GraphicsPathFileNames() As String()
    Get
      GetFileNames()
      Return mFileNames
    End Get
  End Property

  Public Shared ReadOnly Property LastError() As String
    Get
      Return mIOerror
    End Get
  End Property

  Private Shared Sub GetFileNames()
    Dim a As ArrayList = New ArrayList()
    Dim k As Int32

    mIOerror = ""
    Try
      mFileNames = Directory.GetFiles(Directory.GetCurrentDirectory())

      'We are only interested in the graphics path filenames which end in .pth
      For k = 0 To mFileNames.Length - 1
        If Path.GetExtension(mFileNames(k).ToLower()) = ".pth" Then
          a.Add(Path.GetFileName(mFileNames(k)))
        End If
      Next

      'Convert the explicit array to implicit array so I can return string[]
      a.TrimToSize()
      If a.Count > 0 Then
        ReDim mFileNames(a.Count - 1)
        For k = 0 To a.Count - 1
          mFileNames(k) = CType(a(k), String)
        Next
      Else
        ReDim mFileNames(0)
        mFileNames(0) = String.Empty
      End If

    Catch ex As ArgumentNullException
      mIOerror = "Path is a null reference."
    Catch ex As System.Security.SecurityException
      mIOerror = "The caller does not have the required permission."
    Catch ex As ArgumentException
      mIOerror = "Path is an empty string, contains only white spaces, " + _
                 "or contains invalid characters."
    Catch ex As System.IO.DirectoryNotFoundException
      mIOerror = "The path encapsulated in the Directory object does not exist."
    End Try
  End Sub
End Class

File: _REPO/APRESS/gdiplus-programming-in-csharp-vb.net/CH 10/CustomWindow/WindowPath.cs
using System;
using System.Collections;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.IO;

namespace CustomWindow
{
	public sealed class WindowPath
	{
    private static GraphicsPath mP;
    private static string[] mFileNames;
    private static string mIOerror;

    /// <summary>
    /// Gets GraphicsPath stored in file
    /// </summary>
    /// <param name="fname"></param>
    /// <returns GraphicsPath></returns>
    public static GraphicsPath GetPath(string fname)
    {
      Point BeginP = Point.Empty;
      Point EndP   = Point.Empty;
      Point P      = Point.Empty;
      string s;
      string[] x;
      mP = new  GraphicsPath();

      StreamReader sr = new StreamReader(fname);
      sr.BaseStream.Seek(0, SeekOrigin.Begin);
      while (sr.Peek() > -1)
      {
        s = sr.ReadLine();
        x = s.Split(new Char[] {','});
        P.X = Convert.ToInt32(x[0]);
        P.Y = Convert.ToInt32(x[1]);

        if (BeginP == Point.Empty)
        {
          BeginP = P;
          continue;
        }

        EndP = P;
        mP.AddLine( BeginP, EndP);
        BeginP = EndP;
      }
      mP.CloseFigure();

      return mP;
    }

    public static void SavePath(string fname, GraphicsPath P)
    {
      FileStream fs = new FileStream(fname, FileMode.Create);
      StreamWriter sw = new StreamWriter(fs);
      foreach (PointF p in P.PathPoints)
      {
        sw.Write(p.X.ToString() + "," + p.Y.ToString() + "\n");
      }
      sw.Close();
      GetFileNames();
    }

    public static string[] GraphicsPathFileNames
    {
      get 
      { 
        GetFileNames();
        return mFileNames; 
      }
    }

    public static string LastError
    {
      get {return mIOerror;}
    }

    private static void GetFileNames() 
    {
      ArrayList a = new  ArrayList();
      mIOerror = "";
      try 
      {
        mFileNames = Directory.GetFiles(Directory.GetCurrentDirectory()); 

        //We are only interested in the graphics path filenames which end in .pth
        for (int k=0; k<mFileNames.Length; k++)
        {
          if ( Path.GetExtension(mFileNames[k].ToLower()) == ".pth" )
            a.Add(Path.GetFileName(mFileNames[k]));
        }

        //Convert the explicit array to implicit array so I can return string[]
        a.TrimToSize();
        mFileNames = new string[a.Count];
        for (int k=0; k<a.Count; k++)
          mFileNames[k] = (string)a[k];

      }
      catch (ArgumentNullException) 
      {
        mIOerror = "Path is a null reference.";
      }
      catch (System.Security.SecurityException) 
      {
        mIOerror = "The caller does not have the " +
          "required permission.";
      }
      catch (ArgumentException) 
      {
        mIOerror = "Path is an empty string, " +
          "contains only white spaces, " + 
          "or contains invalid characters.";
      }
      catch (System.IO.DirectoryNotFoundException) 
      {
        mIOerror = "The path encapsulated in the " + 
          "Directory object does not exist.";
      }
    }
	}
}
